config/snapshot/config_snapshot.go

```diff
@@ -29,130 +28,67 @@ import (
-//   Each dashboard tabs name -> the configuration of the test group backing it
-func (cs *Config) DashboardTestGroups(dashboardName string) (*configpb.Dashboard, map[string]*configpb.TestGroup) {
-	if cs == nil {
-		return nil, nil
-	}
-	cs.lock.RLock()
-	defer cs.lock.RUnlock()
-
-	d := cs.dashboards[dashboardName]
-	if d == nil {
-		return nil, nil
-	}
-	gs := make(map[string]*configpb.TestGroup, len(d.DashboardTab))
-	for _, tab := range d.DashboardTab {
-		gs[tab.Name] = cs.groups[tab.TestGroupName]
-	}
-	return d, gs
-}
-
-// OnConfigUpdate is a function to be called whenever the config is found to have changed
-// Errors returned by this will be logged, but will not stop the config update.
-type OnConfigUpdate func(*Config) error
-
-// Observe will continuously keep the config up-to-date.
-// If it the config changes, the callback will be called.
-// Runs continuously until the context expires.
-func Observe(ctx context.Context, log logrus.FieldLogger, client gcs.ConditionalClient, configPath gcs.Path, update OnConfigUpdate) (*Config, error) {
-	var cs Config
 	log = log.WithField("observed-path", configPath.String())
 
-	_, err := cs.update(ctx, client, configPath, nil)
+	initialSnap, err := updateHash(ctx, client, configPath, nil)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("can't read %q: %w", configPath.String(), err)
 	}
-	ticker := time.NewTicker(time.Minute) // TODO(fejta): subscribe to notifications
+	var cond storage.Conditions
+	cond.GenerationNotMatch = initialSnap.Attrs.Generation
 
 	go func() {
+		defer close(ch)
+		ch <- initialSnap
+		if ticker == nil {
+			return
+		}
 		for {
-			changed, err := cs.tick(ctx, client, configPath, update)
-			if changed {
-				log.Infof("Configuration changed")
-			}
-			if err != nil {
-				log.Errorf("Error while observing config: %v", err)
-			}
-
 			select {
 			case <-ctx.Done():
-				ticker.Stop()
 				return
-			case <-ticker.C:
+			case <-ticker:
+				snap, err := updateHash(ctx, client, configPath, &cond)
+				switch {
+				case err == nil:
+					// Configuration changed
+					ch <- snap
+					cond.GenerationNotMatch = snap.Attrs.Generation
+				case !gcs.IsPreconditionFailed(err):
+					log.WithError(err).Errorf("Error while observing config")
+					return
+				}
 			}
+
 		}
 	}()
 
-	return &cs, nil
+	return ch, nil
 }
 
-func (cs *Config) tick(ctx context.Context, client gcs.ConditionalClient, configPath gcs.Path, update OnConfigUpdate) (bool, error) {
-	var cond storage.Conditions
-	cond.GenerationNotMatch = cs.attrs.Generation
-	_, err := cs.update(ctx, client, configPath, &cond)
-	switch {
-	case err == nil:
-		// Configuration changed
-		return true, update(cs)
-	case !gcs.IsPreconditionFailed(err):
-		return false, err
+func updateHash(ctx context.Context, client gcs.ConditionalClient, configPath gcs.Path, cond *storage.Conditions) (ConfigSnapshot, error) {
+	cs := ConfigSnapshot{}
```

Comment:
`var cs ConfigSnapshot` -- you're not initializing anything

https://dave.cheney.net/practical-go/presentations/qcon-china.html#_use_a_consistent_declaration_style